### ISSUE TITLE
Allow user set LazyThreadSafetyMode in Delayed Instantianition dependency

### ISSUE THREAD

**Issue by lechu445** [autofac/Autofac#1282]
## Problem Statement
A [delayed instantianition](https://autofac.readthedocs.io/en/latest/resolve/relationships.html#delayed-instantiation-lazy-b) cause locks when multiple threads try to access a given instance (default behavior of `Lazy<T>`). `Lazy<T>` class contains synchronization mechanisms that allows to determine the behavior of threads when creating and retrieving the resource (`LazyThreadSafetyMode` enum).

## Desired Solution
Set `LazyThreadSafetyMode.PublicationOnly` as the default flag when resolving a lazy dependency in Autofac.

## Alternatives You've Considered
Add the ability to set the `LazyThreadSafetyMode` flag during Autofac integration.

**Comment by tillig** [comment]
Something we'd need to determine before making a change like this is the implication on number of expected resolutions, memory usage, perf, etc.

[`PublicationOnly` means](https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazythreadsafetymode?view=net-5.0), "When multiple threads try to initialize a Lazy<T> instance simultaneously, all threads are allowed to run the initialization method (or the parameterless constructor, if there is no initialization method)."

The problem is that, from a general resolution behavior standpoint, **that's not actually expected**.

For example, let's say you have registrations set up like this:

```c#
var b = new ContainerBuilder();
b.RegisterType<Consumer>();
b.RegisterType<Dependency>();
var container = b.Build();
```

And let's say these are very simple types:

```c#
public class Dependency
{
  public void DoWork() => Console.WriteLine("Dep doing work.");
}

public class Consumer
{
  private readonly Dependency _dep;
  public Consumer(Dependency dep)
  {
    this._dep = _dep;
  }
  public void MakeDepDoWork() => this._dep.DoWork();
}
```

If I resolve a Consumer...

```c#
var consumer = container.Resolve<Consumer>();
```

I expect _exactly one_ `Consumer` to be resolved and _exactly one_ `Dependency` to be resolved. Basically, I expect it to be the equivalent of:

```c#
var consumer = new Consumer(new Dependency());
```

In a deferred instantiation, my expectation doesn't change:

```c#
public class Consumer
{
  private readonly Lazy<Dependency> _dep;
  public Consumer(Lazy<Dependency> dep)
  {
    this._dep = _dep;
  }
  public void MakeDepDoWork() => this._dep.Value.DoWork();
}
```

I would expect this, too, to resolve once and only once, though the _time_ at which it resolves is deferred.

The challenge I'd see we'd need to overcome is that **if we _don't_ keep thread synchronization the way it is, we break that expectation**. Per the documentation, it would allow the instantiation logic to run _many times_ even though only one would _set the value_. If you have five threads all doing this at the same time, that'd be like:

```c#
var _ = new Dependency();
var _ = new Dependency();
var _ = new Dependency();
var consumer = new Consumer(new Dependency());
var _ = new Dependency();
```

That is, all five of those threads could _actually resolve_ the thing even though only one is _kept_. The lock in this case is actually very much intentional. Object construction isn't asynchronous and that's what this is - albeit delayed.

If the construction logic in the `Dependency` is complex or resource intensive (you never know, people do all sorts of things), that actually could introduce problems. Maybe you allocate five times the resources but only one gets used. Now you have garbage collection to deal with for the other four objects. It becomes worse if you think that `Dependency` could be a complex object with a lot of dependencies that all have dependencies.

This is possibly less of an issue if you register your object as `SingleInstance()` instead of the default `InstancePerDependency()`, but then you're just pushing the lock down a level - the lifetime scope will still lock to make sure you are properly sharing, so you're not losing the lock so much as moving it.

An alternative might be to use `Func<T>` instead of `Lazy<T>`. Then you could create your own permissive lazy if you don't care about the extra potential resolution operations.

```c#
public class Consumer
{
  public readonly Lazy<Dependency> _dep;
  public Consumer(Func<Dependency> dep)
  {
     _dep = new Lazy<Dependency>(dep, LazyThreadSafetyMode.PublicationOnly);
  }
}
```

**Would that solve the issue?**

I'm really afraid of providing "foot-guns" like this as out-of-the-box features for a small few advanced scenarios simply because of the added support burden. "I set the thread safety mode and now my app is eating double the memory because I didn't think about how the threading was working! Autofac, why did you break me?"

**Comment by tillig** [comment]
Given we've had no input here, there's a workaround, and I'm not really sure this is something we really want to support, I'm going to close this for now.
